internal/rpm/find.go

```diff
@@ -0,0 +1,274 @@
+// Package rpm provides the ability to inspect rpm databases.
+package rpm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"iter"
+	"os"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/internal/rpm/bdb"
+	"github.com/quay/claircore/internal/rpm/ndb"
+	"github.com/quay/claircore/internal/rpm/sqlite"
+)
+
+// FindDBs walks the passed [fs.FS] looking for any files that seem to be rpm
+// databases.
+//
+// The returned [FoundDB] instances close over the passed [fs.FS]
+func FindDBs(ctx context.Context, sys fs.FS) ([]FoundDB, error) {
+	found := make([]FoundDB, 0)
+	if err := fs.WalkDir(sys, ".", walk(ctx, &found, sys)); err != nil {
+		return nil, fmt.Errorf("internal/rpm: error walking fs: %w", err)
+	}
+	slices.SortFunc(found, func(a, b FoundDB) int {
+		cmp := strings.Compare(a.path, b.path)
+		if cmp == 0 {
+			switch {
+			case a.kind < b.kind:
+				cmp = -1
+			case a.kind > b.kind:
+				cmp = +1
+			}
+		}
+		return cmp
+	})
+	found = slices.CompactFunc(found, func(a, b FoundDB) bool {
+		return a.path == b.path
+	})
+	return found, nil
+}
+
+// Walk is a WalkDirFunc to find rpm databases.
+func walk(ctx context.Context, out *[]FoundDB, sys fs.FS) fs.WalkDirFunc {
+	type checkFunc func(context.Context, io.Reader) bool
+	testpath := func(name string, fn checkFunc) (bool, error) {
+		f, err := sys.Open(name)
+		if err != nil {
+			return false, err
+		}
+		ok := fn(ctx, f)
+		return ok, f.Close()
+	}
+
+	return func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		dir, n := path.Split(p)
+		dir = path.Clean(dir)
+		db := FoundDB{
+			sys:  sys,
+			path: dir,
+		}
+		switch n {
+		case `Packages`:
+			db.kind = kindBDB
+			if ok, err := testpath(p, bdb.CheckMagic); !ok {
```

Comment:
confirming we want to ignore `true, err != nil`?

Reply:
Yeah, `testpath` could return that if `fs.File.Close()` fails, I _think_ that case should be accounted for.

Reply:
Failing to close the file should stop the process, IMO.